File: embedding/mvm.rb
require 'java'
java_import 'org.jruby.embed.ScriptingContainer'
java_import 'org.jruby.embed.LocalContextScope'
java_import 'org.jruby.embed.LocalVariableBehavior'

require 'thread'

producer = ScriptingContainer.new(LocalContextScope::SINGLETHREAD, LocalVariableBehavior::PERSISTENT)
queue = Queue.new

producer.put('queue', queue)

producer_thread = Thread.new {
  producer.run_scriptlet(<<-END)
    10.times do
      queue.push('hello')
      sleep 0.1
    end
  END
}

consumer_threads = 10.times.to_enum.map { |idx|
  Thread.new {
    consumer = ScriptingContainer.new(LocalContextScope::SINGLETHREAD, LocalVariableBehavior::PERSISTENT)
    consumer.put('idx', idx)
    consumer.put('queue', queue)
    consumer.run_scriptlet(<<-END)
      while msg = queue.pop
        p [idx, msg]
      end
    END
  }
}

producer_thread.join

# send stop msg to consumers
10.times do
  queue.push(nil)
end

consumer_threads.each(&:join)

File: jjug-jruby-2011/embedding/Mvm.java
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;

import org.jruby.Ruby;
import org.jruby.embed.LocalContextScope;
import org.jruby.embed.LocalVariableBehavior;
import org.jruby.embed.ScriptingContainer;
import org.jruby.runtime.builtin.IRubyObject;

public class Mvm {
    public static void main(String[] args) throws Exception {
        final BlockingQueue<IRubyObject> queue = new LinkedBlockingQueue<IRubyObject>();
        
        Thread producerThread = new Thread(new Runnable() {
            public void run() {
                ScriptingContainer producer = new ScriptingContainer(
                        LocalContextScope.SINGLETHREAD, LocalVariableBehavior.PERSISTENT);
                producer.put("queue", queue);
                producer.runScriptlet(
                    "10.times do\n" +
                      "queue.put('hello')\n" +
                      "sleep 0.1\n" +
                    "end\n"
                );
            }
        });
        
        producerThread.start();

        List<Thread> consumerThreads = new ArrayList<Thread>();
        for (int idx = 0; idx < 10; ++idx) {
            final int thIdx = idx;
            Thread consumerThread = new Thread(new Runnable() {
                public void run() {
                    ScriptingContainer consumer = new ScriptingContainer(
                            LocalContextScope.SINGLETHREAD, LocalVariableBehavior.PERSISTENT);
                    consumer.put("idx", new Integer(thIdx));
                    consumer.put("queue", queue);
                    consumer.runScriptlet(
                        "while msg = queue.take\n" +
                          "p [idx, msg]\n" + 
                        "end\n"
                    );
                }
            });
            consumerThread.start();
            consumerThreads.add(consumerThread);
        }
        
        producerThread.join();

        // send stop msg to consumers
        for (int idx = 0; idx < 10; ++idx) {
            queue.put(Ruby.getGlobalRuntime().getNil());
        }
        for (Thread consumerThread : consumerThreads) {
            consumerThread.join();
        }
    }
}
